### ISSUE TITLE
Matching HashCodes for objects with inner lists

### ISSUE THREAD

**Issue by thomas-bartleet** [felangel/equatable#49]
**Describe the bug**
I have created two objects, both contain a `String` object and a `List<int>` object. The `List<int>` objects are the same, but the `String` object differs. In theory, this should mean the two objects are different. Checking the equality using `==` and `!=` works fine, but the hash codes match exactly.

**To Reproduce**
```dart
void main() {
  test("Equatable check", () {
    final a = _MyClass(items: _createItemList(title: "abcde"));
    final b = _MyClass(items: _createItemList(title: "fghij"));

    expect(a != b, true); // This passes
    final aHashCode = a.hashCode;
    print("a\'s hashcode: $aHashCode");
    final bHashCode = b.hashCode;
    print("b\'s hashcode: $bHashCode");
    expect(aHashCode != bHashCode, true); // This fails.
  });
}

List<ListItem> _createItemList({String title = "mock"}) {
  final result = <ListItem>[
    ListItem(
      title: "title " + title,
      innerList: const <int>[12345],
    ),
  ];
  return result;
}

class _MyClass extends Equatable {
  final List<ListItem> items;

  const _MyClass({
    @required this.items,
  });

  @override
  List<Object> get props => [items];
}

class ListItem extends Equatable {
  final String title;
  final List<int> innerList;

  const ListItem({
    @required this.title,
    this.innerList,
  });

  @override
  List get props => [title, innerList];
}
```

The result is:
```
00:03 +0: Equatable check
a's hashcode: 875940833
b's hashcode: 875940833
00:03 +0 -1: Equatable check [E]
  Expected: <true>
    Actual: <false>
```

However, if I create my `List<ListItem>` without an inner list:
```dart
  final result = <ListItem>[
    ListItem(
      title: "title " + title,
    ),
  ];
```
The hash codes differ and the test passes:
```
00:03 +0: Equatable check
a's hashcode: 237995961
b's hashcode: 268130525
00:03 +1: All tests passed! 
```

**Expected behaviour**
Unless I've misunderstood, I'd expect the `==` operator to be in sync with the hashcode, such that `a == b` is the same as `a.hashcode == b.hashcode`.

**Version**
Running `Equatable 1.0.2`

**Additional context**
I am testing this using Flutter:
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, v1.12.13+hotfix.5, on Mac OS X 10.15.2 19C57, locale en-GB)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 11.3)
[✓] Android Studio (version 3.4)
[✓] VS Code (version 1.41.1)
[✓] Connected device (1 available)

• No issues found!
```

**What I have tried**
- I have tried using the SDK `collection` dependency (1.14.11) and forcing the latest dependency (1.14.12), but both show the same result.
- I have tried creating the lists outside of a method so that:
```dart
const a = _MyClass(items: <ListItem>[
  ListItem(
    title: "ABCDE",
    innerList: <int>[12345],
  ),
]);
const b = _MyClass(items: <ListItem>[
  ListItem(
    title: "FGHIJ",
    innerList: <int>[12345],
  ),
]);
```
- Giving different values in the `innerList` property. **This works** so that the test passes, but I don't understand how two `ListItem`s with the same `innerList` but different `title` properties have the same hashcode:
```dart
final a = _MyClass(items: <ListItem>[
  ListItem(
    title: "ABCDE",
    innerList: <int>[12345],
  ),
]);
final b = _MyClass(items: <ListItem>[
  ListItem(
    title: "FGHIJ",
    innerList: <int>[67890], // Different values cause the test to pass
  ),
]);
```

**Comment by thomas-bartleet** [comment]
Interestingly, using the spread operator (`...`) on the list contents seems to pass the test fine:
```dart
class _MyClass extends Equatable {
  final List<ListItem> items;

  const _MyClass({
    @required this.items,
  });

  @override
  List<Object> get props => [...items];
}

class ListItem extends Equatable {
  final String title;
  final List<int> innerList;

  const ListItem({
    @required this.title,
    this.innerList,
  });

  @override
  List get props => [title, ...innerList];
}
```

Why does this work and not just inserting the list property itself?

**Comment by felangel** [comment]
Hi @thomas-bartleet 👋 
Thanks for opening an issue!

Instances that aren't equal can have the same `hashCode`.

> Objects that are not equal are allowed to have the same hash code, it is even technically allowed that all instances have the same hash code, but if clashes happen too often, it may reduce the efficiency of hash-based data structures like HashSet or HashMap.
https://api.dart.dev/stable/2.7.0/dart-core/Object/hashCode.html

With that being said I'll definitely take a closer look and see if there's anything I can do to eliminate this clash. Thanks for reporting this! 👍

**Comment by felangel** [comment]
I believe this has been resolved by #49 and published in v1.0.2

**Comment by thomas-bartleet** [comment]
Many thanks, @felangel and @pjahn87 !
